### ISSUE TITLE
.has() method unexpected behavior

### ISSUE THREAD

**Issue by Hauts** [kenwheeler/cash#148]
A simplest test (page has divs, of course):
```javascript
console.log( $('body').has('div') );
```
will return $body reference and log 'undefined'.

As in documentation:

> Returns boolean result of the selector argument against the collection.
> 
> ```javascript$(element).has(selector) // => boolean```

Also i have exceptions like

> Uncaught SyntaxError: Failed to execute 'querySelectorAll' on 'Element': '[object Object]' is not a valid selector.

> Uncaught SyntaxError: Failed to execute 'querySelectorAll' on 'Element': '[object HTMLDivElement]' is not a valid selector.

when trying to execute this code:
```javascript
var $self = $('.some-element');

$('body').on('mousedown', function( e ){
    var $target = $(e.target); // or e.target
    if( $self.has( $target ) || $self.is($target) ){
        console.log('Inner click!')
    }
})
```

So this is a critical bug and i have no ideas how to create hotfix to it :/

**Comment by shshaw** [comment]
I believe [$.fn.has()](https://github.com/kenwheeler/cash/blob/master/src/traversal.js#L47-L51) should be returning more like this:

```
  has(selector) {
    return filter.call(this, el => {
      return cash(el).find(selector).length !== 0;
    }).length > 0;
  },
```

I'll attempt some testing and a push this week if I can. In the meantime, `$('body').has('.test').length > 0` should get the result you expect.

**Comment by shshaw** [comment]
The second error seems to be an issue with the `find` method. I thought I'd resolved that, but apparently not...

**Comment by shshaw** [comment]
Scratch my other replies. `$.fn.has` is supposed to return a collection, not a boolean. It filters the collection to only elements that contain that selector. That's the same as jQuery's functionality. If you want to get a boolean from it, `if ( $('body').has('.test').length > 0 ) {` will be what you need.

I have a pull request in the queue ( #133 ) that fixes some issues with `has` and fixes the second error as well, I just haven't merged that yet.

**Comment by Hauts** [comment]
@shshaw I think the major problem is that core [find](https://github.com/kenwheeler/cash/blob/master/src/core.js#L10) method does not work with any Cash collections. Because of it it is impossible to pass Cash collection as selector and this is a really poor side of library.

About boolean / collection returns by .has — that's no matter for me :) All i need is to have functionality to pass collections as a selector and get expected results.

P.S. Thanks for response and sorry for my english.

**Comment by shshaw** [comment]
For your example, you could simply change this to get the functionality wanted.
```
$('body').on('mousedown', function( e ){
    if( $self.has( e.target ) || $self.is(e.target) ){
    ...
});
```

Yes, in general, better collection support has been something I've been adding where I can. It has to be managed carefully to keep performance up, so it's more than just wrapping everything as a collection and running `.each`

**Comment by Hauts** [comment]
I've tried that case (with e.target) and get exception. I will test tomorrow one more time with your PR.

And my example is really simple. In real case it e.target can be a large collection of elements from lots of selectors, so it would be hard to deconstruct

$self.has( $someLargeCollection )

to

$self.has( '.some-class-1' ) || $self.has( 'some-class-2' ) ... $self.has( 'some-class-N' )

So.. I will wait to use Cash before collection functionality will works :)

**Comment by shshaw** [comment]
Sure, I understand the desire for collection support, I was just referencing your specific example where you're only checking one element, the target of the event.

For your second example, you can do multiple string selectors with `$.fn.has`, so if you're not actually referencing a specific element and just checking for classes, then `$self.has('.some-class-1, .some-class-2, .some-class-3')` will work.

Just passing along some ideas, not dismissing the collection support request.

**Comment by vivekimsit** [comment]
@shshaw this should be closed as per #133

**Comment by shshaw** [comment]
I was leaving it open for the "collection support" side of it, but I suppose that should be a separate issue to keep this cleaner.

Thanks, @vivekimsit
